src/Nethermind/Nethermind.Runner/Ethereum/Steps/StartBlockProducer.cs

```diff
@@ -66,11 +69,10 @@ protected virtual void BuildProducer()
 
         protected BlockProducerContext GetProducerChain()
         {
-            ReadOnlyDbProvider readOnlyDbProvider = new ReadOnlyDbProvider(_context.DbProvider, false);
             ReadOnlyBlockTree readOnlyBlockTree = new ReadOnlyBlockTree(_context.BlockTree);
-            ReadOnlyTxProcessingEnv readOnlyTxProcessingEnv = new ReadOnlyTxProcessingEnv(readOnlyDbProvider, readOnlyBlockTree, _context.SpecProvider, _context.LogManager);
-            BlockProcessor blockProcessor = CreateBlockProcessor(readOnlyTxProcessingEnv, readOnlyDbProvider);
-            OneTimeChainProcessor chainProcessor = new OneTimeChainProcessor(readOnlyDbProvider, new BlockchainProcessor(readOnlyBlockTree, blockProcessor, _context.RecoveryStep, _context.LogManager, false));
+            ReadOnlyTxProcessingEnv readOnlyTxProcessingEnv = new ReadOnlyTxProcessingEnv(_readOnlyDbProvider, readOnlyBlockTree, _context.SpecProvider, _context.LogManager);
+            BlockProcessor blockProcessor = CreateBlockProcessor(readOnlyTxProcessingEnv, _readOnlyDbProvider);
+            OneTimeChainProcessor chainProcessor = new OneTimeChainProcessor(_readOnlyDbProvider, new BlockchainProcessor(readOnlyBlockTree, blockProcessor, _context.RecoveryStep, _context.LogManager, false));
             var pendingTxSelector = CreatePendingTxSelector();
```

Comment:
pass ReadOnlyDbProvider as a parameter to CreatePendingTxSelector instead of class field

Reply:
but then you impose on each implementation to require db provider, not method seems perfect to me here